### ISSUE TITLE
cgroup-path

### ISSUE THREAD

**Issue by ansd** [cloudfoundry-incubator/bpm-release#109]
**TLDR:**
1. prepend `bpm` to the cgroup path
2. do not base32 encode the container id

**Details:**

In a [previous story](https://www.pivotaltracker.com/n/projects/2070399/stories/150819404), the decision was to base32 encode the job name for the runc container-id which becomes the cgroup-path in the file `/proc/[pid]/cgroup`:
```
hierarchy-ID:controller-list:cgroup-path
``` 

I would like to write a [sysdig-bosh release](https://github.com/ansd/sysdig-boshrelease) which will be colocated on VMs and which should ideally also be able to monitor bpm containers. Sysdig parses the cgroup-path in order to classify the container type (see [here](https://github.com/draios/sysdig/blob/f27e1ca1f17cbe6a463b7e248867c488f49fecf4/userspace/libsinsp/container.cpp#L581) for LXC).
For example, the cgroup-path for an LXC container looks like `/lxc/my-container`:
```
cat /proc/<pid of lxc container>/cgroup
11:hugetlb:/lxc/my-container
```

It's the default for LXC (see [here](https://linuxcontainers.org/lxc/manpages/man5/lxc.system.conf.5.html)):
```
lxc.cgroup.pattern
Format string used to generate the cgroup path (e.g. lxc/%n).
```

According to the sysdig code, the same applies to [mesos](https://github.com/draios/sysdig/blob/f27e1ca1f17cbe6a463b7e248867c488f49fecf4/userspace/libsinsp/container.cpp#L684), [libvirt-lxc](https://github.com/draios/sysdig/blob/f27e1ca1f17cbe6a463b7e248867c488f49fecf4/userspace/libsinsp/container.cpp#L616) and [docker](https://github.com/draios/sysdig/blob/f27e1ca1f17cbe6a463b7e248867c488f49fecf4/userspace/libsinsp/container.cpp#L480).
Only the bpm container has this base32 encoded cgroup-path like
```
12:pids:/MJWG6YTTORXXEZI-
```
for the blobstore for example.

Having prepended something like `bpm.` to the cgroup-path would allow to parse it easily in sysdig such that the filter `csysdig -pc container.type=bpm` could match BPM containers.

But even though if the cgroup-path looks like `bpm.MJWG6YTTORXXEZI-`, the output for above filter would show a container column with the name `MJWG6YTTORXXEZI-` rather than `blobstore`.

Therefore, it would also be better to not base32 encode the whole container name, but encode maybe only those chars which do not match the regex `^[\w+-\.]+$` (see [code](https://github.com/opencontainers/runc/blob/079817cc26ec5292ac375bb9f47f373d33574949/libcontainer/factory_linux.go#L32)). This would probably require to have a special delimiter char for the decoding process in order to identify the encoded chars.

See also [Slack discussion](https://cloudfoundry.slack.com/archives/C7A0K6NMU/p1541758095018100) for more context.

If the above two requirements are okay with you, I could also create a PR.

**Comment by xoebus** [comment]
Just merged a new encoding format for BPM job IDs. It should go out in the next release. Let us know if it's missing anything.

Thanks for the issue!

**Comment by ansd** [comment]
@xoebus many thanks for changing the container-id.

As a side note:
After testing your change I realise that the cgroup-path is not just the container-id anymore:
```
bosh/0:~# cat /proc/<blobstore pid>/cgroup
12:devices:/system.slice/runc-bpm-blobstore.scope
```
So, the cgroup-path is `/system.slice/runc-bpm-blobstore.scope` rather than just `bpm-blobstore`.
This was introduced by [pass --systemd-cgroup when on a systemd machine](https://github.com/cloudfoundry-incubator/bpm-release/commit/882c587bafa2b7cdd2bfbc89df60091b27bb3d61).
`/system.slice/runc-` comes from [here](https://github.com/opencontainers/runc/blob/751f18de2af90495e9c5665b95bfc7adf66ddd57/libcontainer/specconv/spec_linux.go#L318-L322) and `.scope` comes from [here](https://github.com/opencontainers/runc/blob/751f18de2af90495e9c5665b95bfc7adf66ddd57/libcontainer/cgroups/systemd/apply_systemd.go#L528).

(The `parent` (`system.slice`), `ScopePrefix` (`runc`), and the appended `scope` (vs `slice`) are configurable.)

It makes parsing the cgroup-path in sysdig less straightforward.
However, I think that parsing the cgroup-path for `bpm-` on systemd and non-systemd machines still identifies the BPM container type. And the container-id is then the string between `bpm-` and either `.scope` or the end of the string.

**Comment by xoebus** [comment]
Changing the prefix will change behavior as it's changing the cgroup hierarchy. We're going to probably need to change that soon as we start doing CPU share limiting which is very dependent on the hierarchy. I don't know if we can make any guarantees about what that path will be.

**Comment by ansd** [comment]
@xoebus hmm I'm not that deep into how CPU share limiting works and how the cgroup hierachy gets affected, and  I don't know whether the following suggestion makes sense but wouldn't it be possible that the cgroup-path stays somewhat stable and BPM identifiable by introducing a bpm root cgroup like Mesos is doing it for example (see https://mesos.readthedocs.io/en/latest/configuration/):
```
--cgroups_root=VALUE	Name of the root cgroup. (default: mesos)
```

LXC seems to also have this root cgroup (see https://linuxcontainers.org/lxc/manpages/man5/lxc.system.conf.5.html):
```
lxc.cgroup.pattern
Format string used to generate the cgroup path (e.g. lxc/%n).
```

That way you could introduce any other hierarchy (including CPU share limiting) below this root cgroup.
